### ISSUE TITLE
draw_networkx_nodes misshandles ax (easy fix).

### ISSUE THREAD

**Issue by PWills2013** [networkx/networkx#3443]
Symtoms:
    Causes clipped nodes and non-transparent background.  

Recreate issue:
```
import pandas as pd
import numpy as np
import networkx as nx
import matplotlib.pyplot as plt
 
# Build a dataframe with your connections
df = pd.DataFrame({ 'from':['A', 'B', 'C','A'], 'to':['D', 'A', 'E','C'] })
df
 
# Build your graph
G=nx.from_pandas_edgelist(df, 'from', 'to', create_using=nx.Graph() )
 
# Custom the nodes:
fig = plt.figure()
pos = nx.spectral_layout(G)
nx.draw_networkx_nodes(G, pos=pos, with_labels=True, node_color='skyblue', node_size=1500, edge_color='white')
fig.set_facecolor("#00000F")

plt.show()
```

In `networkx/drawing/nx_pylab.py` the `draw_networkx_nodes` function 

currently:
```python
if ax is None:
    ax = plt.gca()
```

should be something like:
```python
if ax is None:
    cf = plt.gcf()
    if cf._axstack() is None:
        ax = cf.add_axes((0, 0, 1, 1))
    else:
        ax = cf.gca()
else:
    cf = ax.get_figure()
cf.set_facecolor('w')
ax.set_axis_off()
```

Workaround:
    Add "ax == None" logic from the `should be something like` section above and pass the ax key word arg to `draw_networkx_nodes`

*edit:  added complete example of issue.

**Comment by dschult** [comment]
Thanks for this!

I'm not sure I understand the difference. I understand that with your example code the nodes are "too big" and so they get clipped at the edge of the figure. What is the picture supposed to look like with the improved call to ax? In my version it seems like the nodes are still clipped.

Also, why do you need ```set_facecolor('w')```? 
Isn't that limiting if someone wants a different color?

Does setting the axis off affect the rest of this or just to make the example clearer?

**Comment by Gopal-Dahale** [comment]
Hi @MridulS!
I would like to work on this issue, to get acquainted with the networkx for Gsoc.
Please let me know if this is a reasonable initial attempt. Thank you.

**Comment by rossbar** [comment]
Node clipping is a common problem with `nx_pylab` - the simplest workaround which is employed in our example gallery quite often, is to use `ax.margins` to increase the padding on the axis borders.

Here's a more minimal version of the example from the OP (note that the original example doesn't work, because some of the kwargs used in `draw_networkx_nodes` aren't recognized):

```python
>>> import networkx as nx
>>> G = nx.Graph([('A', 'D'), ('B', 'A'), ('C', 'E'), ('A', 'C')])
>>> fig = plt.figure()
>>> pos = nx.spectral_layout(G)
>>> nx.draw_networkx_nodes(G, pos=pos, node_color='skyblue', node_size=1500)
>>> nx.draw_networkx_labels(G, pos)
```

Which gives:

![Figure_1](https://user-images.githubusercontent.com/1268991/112763679-4dfc4700-8fba-11eb-91a9-50817558199d.png)

... where the node-clipping issue is clear. The clipping can be addressed by increasing the padding on the axis margins:

```python
>>> ax = fig.get_axes()[0]
>>> ax.margins(0.1, 0.1)
```

![Figure_2](https://user-images.githubusercontent.com/1268991/112763841-00340e80-8fbb-11eb-97bf-b9a182fc140e.png)

I agree with @dschult that the proposed changes in the OP don't quite address the issue. If I were going to try to tackle the node clipping problem, I'd start by investigating how matplotlib treats markersize during axis creation/resizing, and see if there isn't a general way that `nx_pylab` could account for the full marker extent during drawing.

**Comment by EricPostMaster** [comment]
Jumping in because this seems like an interesting problem, if not a bit sticky...

@rossbar, I looked into how matplotlib treats markersize. I was hoping that we could just use `ax.relim()` to recompute the data limits, taking marker size into account. Unfortunately, the `draw_networkx_nodes` function returns a `matplotlib.collections.PathCollection` object, and `collections` objects are not currently supported by the `ax.relim()` method. So, it seems like matplotlib doesn't take marker size into account when drawing the plot.

So far, the best thing I've been able to come up with is changing the `clip_on` parameter to False in the `node_collection` section of the `draw_networkx_nodes` function, but that just gets the markers to extend past the border of the figure (image below).

```python
node_collection = ax.scatter(
        xy[:, 0],
        xy[:, 1],
        s=node_size,
        c=node_color,
        marker=node_shape,
        cmap=cmap,
        vmin=vmin,
        vmax=vmax,
        alpha=alpha,
        linewidths=linewidths,
        edgecolors=edgecolors,
        label=label,
        clip_on=False,
    )
```



I had a couple of ideas, both of which may be sub-optimal, so I'd love some feedback:

- Turn off the border of the figure by default. If there's no visible border and the markers are not cut off, then the only place they might get cut off is if the plot is big enough to go beyond the size of the entire figure (see image below).

![image](https://user-images.githubusercontent.com/58200144/115965872-29cc5100-a4f9-11eb-83be-877f554303cf.png)

- Adjust `ax.margins()` at the end of the `draw_networkx_nodes` function to account for some percentage of the node size, like maybe half or something. I'm still a bit confused as to how to do that because `ax.margins()` is measured in inches, but `node_size` is in points. Google says points to inches is a 72:1 ratio
    - Potential drawback to this: It might create some goofy scale problems. If an array of node sizes was passed, we'd have to build in something to use the average, largest, smallest, etc.
    - A related possibility would be to take the size of the node and then add it to the axis lengths.

@Gopal-Dahale, have you been working on this at all?

**Comment by EricPostMaster** [comment]
Okay, one more quick thing I just tried... I added this to the end of the `draw_networkx_nodes` function to set `ax.margins()` based on the provided `node_size` parameter.

```python
    # Currently only works if a single node size is passed
    if node_size >= 500:
        ax.margins((node_size/100)/72,(node_size/100)/72)
    else:
        #small markers get a bit more slack
        ax.margins((node_size/50)/72,(node_size/50)/72)
```

### Results:
#### `node_size=1100`
![image](https://user-images.githubusercontent.com/58200144/115966669-add40800-a4fc-11eb-8964-35d4f19b4d70.png)

#### `node_size=800`
![image](https://user-images.githubusercontent.com/58200144/115966607-6a799980-a4fc-11eb-8381-b97265885b6e.png)

#### `node_size=500`
![image](https://user-images.githubusercontent.com/58200144/115966638-93019380-a4fc-11eb-9947-1bacc9b2789e.png)

#### `node_size=300` - `else` statement has kicked in
![image](https://user-images.githubusercontent.com/58200144/115966649-9bf26500-a4fc-11eb-8a1b-abd271ab3ae7.png)

Not sure if it's necessary, but if we can get the coordinates of each of the points, we could identify the node sizes of the furthest node vertically and the furthest node horizontally and then set the margins based on them. Could be done using the `pos` parameter, but I'm not sure how to connect that back to the corresponding node size.

**Comment by Gopal-Dahale** [comment]
> @Gopal-Dahale, have you been working on this at all?
Not now. You may work on this if would like to.

**Comment by EricPostMaster** [comment]
@Gopal-Dahale - I wouldn't mind the help if you want to take a look at it as well! Matplotlib can really give me fits, so I'd love to have an extra pair of eyes. 👀 😄 If not, that's fine too - I just wanted to make sure you knew I wasn't trying to keep it for myself!

**Comment by dschult** [comment]
The order of the nodes in `pos` and `node_sizes should be the same. So, perhaps something like this can parse the position and size together (untested).
```python
for npos, nsize in zip(pos, node_size):
    ...
```

**Comment by rossbar** [comment]
https://github.com/networkx/networkx/issues/3443#issuecomment-826122531

This is a nice solution, but I too would be worried about strange re-scaling problems in corner cases. Ideally this is something that would be handled upstream in Matplotlib (or through the use of mpl API) without requiring any custom axis-handling code to `nx_pylab`.

That said, this issue does seem to affect a lot of users. I think I would still prefer increasing the default axis padding in `nx_pylab`, say `ax.margins(0.15, 0.15)` which would help with some cases. The tradeoffs here are:
 - `-` doesn't actually "fix" the problem, just avoids it for a larger set of cases
 - `-` may result in too much edge space for images with very small nodes
 - `+` avoids adding custom scaling in NX

**Comment by EricPostMaster** [comment]
I've continued researching a bit today and haven't been able to turn up anything in Matplotlib that takes marker size into account when drawing the plot. (maddening!) The closest thing I found is a [bubble chart example](https://matplotlib.org/stable/gallery/misc/packed_bubbles.html#sphx-glr-gallery-misc-packed-bubbles-py), but I think the bubbles are treated differently because they are areas rather than markers.

@rossbar, I thought more about edge cases that might cause re-scaling issues, particularly if one of your edge nodes happens to be relatively enormous and squishes everything to the middle of the plot. The simplest solution I've come up with so far is a simple `auto_margins` parameter for the function that, when `True`, uses the code below to re-calculate margins. Otherwise, it just goes with the NX defaults. The default could be set to whichever you think is more commonly needed.

Thanks, @dschult, for the `zip()` suggestion. I used that to create x- and y-axis margins based on the largest nodes closest to the edges of the plot. I forked the NX repo and created my own [branch with the changes](https://github.com/EricPostMaster/networkx/blob/branch_sims/networkx/drawing/nx_pylab.py) to `nx_pylab`, but this is the main piece I added to the function:

```python
    if auto_margins:
        node_coords_sizes = list(zip(pos.items(), node_size_list))

        x_min = sorted(node_coords_sizes, key=lambda a:a[0][1][0], reverse=False)[0]
        x_max = sorted(node_coords_sizes, key=lambda a:a[0][1][0], reverse=True)[0]
        x_node_size = max([x_min, x_max], key=itemgetter(1))[1]

        if x_node_size >= 500:
            x_margin = (x_node_size/100)/72
        else:
            x_margin = (x_node_size/50)/72

        y_min = sorted(node_coords_sizes, key=lambda a:a[0][1][1], reverse=False)[0]
        y_max = sorted(node_coords_sizes, key=lambda a:a[0][1][1], reverse=True)[0]
        y_node_size = max([y_min, y_max], key=itemgetter(1))[1]

        if y_node_size >= 500:
            y_margin = (y_node_size/100)/72
        else:
            y_margin = (y_node_size/50)/72

        ax.margins(x_margin, y_margin)
```

It takes the nodes that are furthest to the right (`x_max`) and left (`x_min`), selects the largest of the two (`x_node_size`), and then calculates the new horizontal margin (`x_margin`). Then it does the same thing with the highest and lowest points and calculates the new vertical margin. Lastly, `ax.margins()` is assigned with those two values.

I'd love feedback on how to get this to the point where it could be merged. Specifically, there's probably a straightforward way to DRY this out, but I'm just not seeing it yet.

**Comment by stefanv** [comment]
Instead of scatter, could a person use a PatchCollection of circles?  See e.g. [this S/O answer](https://stackoverflow.com/a/48174228/214686), which also shows how to compute point size in terms of viewport size.

**Comment by EricPostMaster** [comment]
@stefanv, PatchCollections are new to me. The S/O page you linked only shows points of one size - do you know if they can plot points of varying sizes? Also, do you know if changing the `ax.scatter` here might impact how the related edge/label functions might display?

**Comment by dschult** [comment]
Maybe you could turn your branch with that solution into a Pull Request. Then we can see how the tests work with it and maybe play and add more tests to it, to make sure it is DRY before we merge. :}
We'll have to keep track of processing time too. I don't think your example code would add too much.
The Patch Collections might be faster or slower and its hard for me to predict.

**Comment by EricPostMaster** [comment]
Cool, I'll get on it. I probably won't get to it tonight, but should hopefully be able to in the next 24 hours 👍

**Comment by EricPostMaster** [comment]
Done! Funny how fast "the next 24 hours" turns into 72 hours 😅

**Comment by dschult** [comment]
Closed by #4769
